File: trunk/C++ Crossplatform/OPSCrossPlatformCppLib/source/OPSObjectFactory.cpp
#include "OPSObjectFactory.h"

#include "Topic.h"
#include "DefaultOPSConfigImpl.h"
#include "MulticastDomain.h"

#include <boost/algorithm/string/split.hpp> 
#include <boost/algorithm/string/Classification.hpp>

namespace ops
{

class BuiltInFactory : public SerializableFactory
{

public:
	Serializable* create(std::string& type)
    {
        if(type == ("ops.protocol.OPSMessage"))
        {
            return new OPSMessage();
        }
		if(type == ("Topic"))
        {
            return new Topic();
        }
		if(type == ("DefaultOPSConfigImpl"))
        {
            return new DefaultOPSConfigImpl();
        }
		if(type == ("MulticastDomain"))
        {
            return new MulticastDomain();
        }
        return NULL;
    }
};

class OPSObjectFactoryImpl : public OPSObjectFactory
{
    
public:
    
    /**
     * Tries to construct the most specialized version of the given typeString
     */
	Serializable* create(std::string& typeString)
    {
		std::vector<std::string> types;
		boost::algorithm::split(types, typeString, boost::algorithm::is_any_of(" "));

        for(int i = 0; i < types.size(); i++)
        {
			Serializable* serializable = SerializableCompositeFactory::create(types[i]);
            if(serializable != NULL)
            {
                return serializable;
            }
        }

        return NULL;
    }
	~OPSObjectFactoryImpl()
	{

	}

};


	/**
     * Create singelton instance of OPSObjectFactory.
     * @return
     */
	//static
    OPSObjectFactory* OPSObjectFactory::getInstance()
    {
		static OPSObjectFactory* instance = NULL;
        if(instance == NULL)
        {
            instance = new OPSObjectFactoryImpl();
            instance->add(new BuiltInFactory());
        }
        return instance;
    }
	

}

File: Java/OPSJLib/src/ops/archiver/OPSObjectFactory.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops.archiver;

import configlib.Serializable;
import configlib.SerializableCompositeFactory;
import configlib.SerializableFactory;
import ops.DefaultOPSConfigImpl;
import ops.Domain;
import ops.Topic;
import ops.protocol.OPSMessage;
import ops.ParticipantInfoData;
import ops.TopicInfoData;

/**
 *
 * @author angr
 */
public class OPSObjectFactory extends SerializableCompositeFactory
{
    static OPSObjectFactory instance = null;
    /**
     * Create singelton instance of OPSObjectFactory.
     * @return
     */
    public static OPSObjectFactory getInstance()
    {
        if(instance == null)
        {
            instance = new OPSObjectFactory();
            instance.add(new SerializableFactory() {

                public Serializable create(String type)
                {
                    if(type.equals("ops.protocol.OPSMessage"))
                    {
                        return new OPSMessage();
                    }
                    if(type.equals("DefaultOPSConfigImpl"))
                    {
                        return new DefaultOPSConfigImpl();
                    }
                    if(type.equals("Domain"))
                    {
                        return new Domain();
                    }
                    if(type.equals("MulticastDomain"))
                    {
                        return new Domain();
                    }
                    if(type.equals("Topic"))
                    {
                        return new Topic();
                    }
                    if (type.equals("ops.ParticipantInfoData"))
                    {
                        return new ParticipantInfoData();
                    }
                    if (type.equals("TopicInfoData"))
                    {
                        return new TopicInfoData();
                    }

                    return null;
                }
            });
        }
        return instance;
    }
    /**
     * Tries to construct the most specialized version of the given typeString
     */
    @Override
    public Serializable create(String typeString)
    {
        String[] types = typeString.split(" ");
        for (String type : types)
        {
            Serializable serializable = super.create(type);
            if(serializable != null)
            {
                return serializable;
            }
        }

        return null;
    }



}
